examples/pennylane_run_quantum_natural_gradient.py

```diff
@@ -0,0 +1,401 @@
+#
+# Consider a variational quantum circuit
+#
+# .. math::
+#
+#     U(\mathbf{\theta})|\psi_0\rangle = V_L(\theta_L) W_L V_{L-1}(\theta_{L-1}) W_{L-1}
+#       \cdots V_{\ell}(\theta_{\ell}) W_{\ell} \cdots V_{0}(\theta_{0}) W_{0} |\psi_0\rangle
+#
+# where
+#
+# * :math:`|\psi_0\rangle` is the initial state,
+# * :math:`W_\ell` are layers of of non-parametrized quantum gates,
+# * :math:`V_\ell(\theta_\ell)` are layers of of parametrized quantum gates
+#   with :math:`n` parameters :math:`\theta_\ell = \{\theta^{(\ell)}_0, \dots, \theta^{(\ell)}_n\}`.
+#
+# Further, assume all parametrized gates can be written in the form
+# :math:`X(\theta^{(\ell)}_{i}) = e^{i\theta^{(\ell)}_{i} K_i}`,
+# where :math:`K^{(\ell)}_i` is the *generator* of the parametrized operation.
+#
+# For each parametric layer :math:`\ell` in the variational quantum circuit
+# the :math:`n_\ell\times n_\ell` block-diagonal submatrix
+# of the Fubini-Study tensor :math:`g_{ij}^{(\ell)}` is calculated by:
+#
+# .. math::
+#
+#     g_{ij}^{(\ell)} = \langle \psi_{\ell-1} | K_i K_j | \psi_{\ell-1} \rangle
+#     - \langle \psi_{\ell-1} | K_i | \psi_{\ell-1}\rangle
+#     \langle \psi_{\ell-1} |K_j | \psi_{\ell-1}\rangle
+#
+# where
+#
+# .. math::
+#
+#     | \psi_{\ell-1}\rangle = V_{\ell-1}(\theta_{\ell-1}) W_{\ell-1} \cdots V_{0}(\theta_{0}) W_{0} |\psi_0\rangle.
+#
+# (that is, :math:`|\psi_{\ell-1}\rangle` is the quantum state prior to the application
+# of parameterized layer :math:`\ell`), and we have :math:`K_i \equiv K_i^{(\ell)}` for brevity.
+#
+# Let's consider a small variational quantum circuit example coded in PennyLane:
+import numpy as np
+
+import pennylane as qml
+from pennylane import expval, var
+
+dev = qml.device('default.qubit', wires=3)
+
+@qml.qnode(dev)
+def circuit(params):
+    # non-parametrized gates
+    qml.RY(np.pi/4, wires=0)
+    qml.RY(np.pi/3, wires=1)
+    qml.RY(np.pi/7, wires=2)
+
+    # Parametrized layer 1
+    qml.RZ(params[0], wires=0)
+    qml.RZ(params[1], wires=1)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    # Parametrized layer 2
+    qml.RY(params[2], wires=1)
+    qml.RX(params[3], wires=2)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    return qml.expval(qml.PauliY(0))
+
+params = np.array([0.432, -0.123, 0.543, 0.233])
+
+##############################################################################
+# The above circuit consists of 4 parameters, with two distinct parametrized
+# layers of 2 parameters each. Therefore, the block-diagonal approximation
+# consists of two :math:`2\times 2` matrices, :math:`g^{(1)}` and :math:`g^{(2)}`.
+#
+# Computing the first block-diagonal :math:`g^{(1)}`, we create subcircuits consisting
+# of all gates prior to the layer, and observables corresponding to
+# the *generators* of the gates in the layer:
+
+g1 = np.zeros([2, 2])
+
+def layer1_subcircuit(params):

```

Comment:
This can have zero arguments, no? params is not used in the function

Reply:
True, I kept it here to be consistent :thinking: Should I remove it or leave it as is?

Reply:
I dont think programming functions need this consistency.  :)